[Docs] Add Guide to Write C++ TM (#4254)
facebook/react-native-website: docs/the-new-architecture/pure-cxx-modules.md

# Pure C++ Turbo Native Modules

import Tabs from '@theme/Tabs'; import TabItem from '@theme/TabItem'; import constants from '@site/core/TabsConstants';

> Writing a module in C++ is the best way to share platform-agnostic code between Android and iOS. With pure C++ modules, you can write your logic only once and reuse it right away from all the platform, without the need of writing platform specific code.

Author: This is not advanced to me. This is a base case of a user that needs to write a Native Module and wants share code between platforms. I'm afraid that if we put the "advanced" tag here, it will actually discourage people to try them out.

The other guide has `Advanced` in the title, instead.

In this guide, we will go through the creation of a pure C++ Turbo Native Module:

1. Create the JS specs
2. Configure Codegen to generate the scaffolding
3. Implement the Native logic
4. Register the module in the Android and iOS application
5. Test your changes in JS

The rest of this guide assume that you have created your application running the command:

```shell
npx @react-native-community/cli@latest init SampleApp --version 0.76.0
```

## 1. Create the JS specs

> Pure C++ Turbo Native Modules are Turbo Native Modules. They needs a specification file (also called spec file) so that Codegen can create the scaffolding code for us. The Specification file is also what we use to access the Turbo Native Module in JS.

Reviewer: ```suggestion
Pure C++ Turbo Native Modules are Turbo Native Modules. They needs a specification file (also called spec file) so that Codegen can create the scaffolding code for us. The specification file is also what we use to access the Turbo Native Module in JS.
```

Specs files need to be written in a typed JS dialect. React Native currently supports Flow or TypeScript

1. Inside the root folder of your app, create a new folder called `specs`.
2. Create a new file called `NativeSampleModule.ts`

> :::warning
> All Native Turbo Module spec files must have the prefix `Native`. Otherwise Codegen cannot pick them up.
> :::

Reviewer: ```suggestion
:::warning
All Native Turbo Module spec files must have the prefix `Native`, otherwise Codegen will ignore them.
:::
```

<Tabs groupId="tnm-specs" queryString defaultValue={constants.defaultJavaScriptSpecLanguages} values={constants.javaScriptSpecLanguages}>
<TabItem value="flow">

```ts
// @flow
import type {TurboModule} from 'react-native'
import { TurboModuleRegistry } from "react-native";

export interface Spec extends TurboModule {
  +reverseString: (input: string) => string;
}

export default (TurboModuleRegistry.getEnforcing<Spec>(
  "NativeSampleModule"
): Spec);
```

</TabItem>
<TabItem value="typescript">

```ts
import {TurboModule, TurboModuleRegistry} from 'react-native';

export interface Spec extends TurboModule {
  readonly reverseString: (input: string) => string;
}

export default TurboModuleRegistry.getEnforcing<Spec>(
  'NativeSampleModule',
);
```

</TabItem>
</Tabs>

## 2. Configure Codegen

> The next step is to configure Codegen in your `package.json`. <!-- Add links to Codegen -->

Reviewer: TODO here

1. Open the `package.json` file
2. Modify it as it follows:

```diff
{
  "name": "SampleApp",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "android": "react-native run-android",
    "ios": "react-native run-ios",
    "lint": "eslint .",
    "start": "react-native start",
    "test": "jest"
  },
  "dependencies": {
    "react": "18.3.1",

>     "react-native": "0.76.0-rc.3"

Reviewer: Can we change this to 0.76.0?

The module lives in the `facebook::react` namespace to have access to all the types that live in that namespace.

The class `NativeSampleModule` is the actual Turbo Native Mdoule class and it extends the `NativeSampleModuleCxxSpec` class which contains some glue code and boilerplate code to let this class behave as a Turbo Native Module.

Finally, we have the constructor, that accepts a pointer to the `CallInvoker`, to communicate with JS if needed and the function's prototype we have to implement.

The `NativeSampleModule.cpp` files is the actual implementation of our Turbo Native Module and implements the constructor and the method that we declared in the specs.

## 4. Register the Module in the platform

The next steps will let us register the module in the platform. This is the step that exposes the native code to JS so that the React Native application can finally call the native methods from the JS layer.

This is the only time when we will have to write some platform-specific code.

### Android

To make sure that the iOS app can effectively build the C++ Turbo Native Module, we need to:

1. Create a `CMakeLists.txt` to access our C++ code

> 2. Modify `gradle.build` to point to the newly created `CMakeLists.txt` file.

Reviewer: ```suggestion
2. Modify `build.gradle` to point to the newly created `CMakeLists.txt` file.
```

> 3. Create an `onLoad.cpp` file in our Android app to register the new Turbo Native Module.

Reviewer: ```suggestion
3. Create an `OnLoad.cpp` file in our Android app to register the new Turbo Native Module.
```

> #### 1. Create the `CMakeList` file

Reviewer: ```suggestion
#### 1. Create the `CMakeLists.txt` file
```

Android uses CMake to build. CMake needs to access the files we defined in our shared folder, to be able to build them.

> 1. Create a new folder `SampleApp/android/app/src/main/jni`. The `jni` folder is where the C++ side of Android lives

Reviewer: Does your path always start with `SampleApp/`? I would rather start with `./` here

Author: The guide starts with a command that creates the `SampleApp` using the CLI. I feel that repeating `SampleApp` gives the user an anchor to understand where to start looking. otherwise `./android` can be ambiguous and might raise questions on what is the starting point.
But I'm okay to change it, if we disagree.

Reviewer: No strong opinion here as long as we're consistent

Author: yep, looks like we are consistent. I check all the usages of `/`

2. Create a `CMakeLists.txt` file and add this context

```shell
cmake_minimum_required(VERSION 3.13)

# Define the library name here.
project(appmodules)

> # This file includes all the necessary to let you build your application with the New Architecture.

Reviewer: ```suggestion
# This file includes all the necessary to let you build your React Native application
```

include(${REACT_ANDROID_DIR}/cmake-utils/ReactNative-application.cmake)

# Define where the additional source code lives. We need to crawl back the jni, main, src, app, android folders
target_sources(${CMAKE_PROJECT_NAME} PRIVATE ../../../../../shared/NativeSampleModule.cpp)

# Define where CMake can find the additional header files. We need to crawl back the jni, main, src, app, android folders
target_include_directories(${CMAKE_PROJECT_NAME} PUBLIC ../../../../../shared)
```

The CMake file does the following things:

> - Specifies the version of CMake we need to use and sets the `VERBOSE` flag.

Reviewer: You can remove this (also it does not set the VERBOSE flag)

> - Defines the library name to which our code will be added
> - Includes the base React Native's CMake file
> - Adds the source code that we need to build with the `target_sources` directives. You can see that we need to crawl back from the `jni` folder to the `shared` folder where our C++ TM lives.
> - Specify where CMake can find the additional header files. Also in this case we need to crawl back from the `jni` folder.

Reviewer: ```suggestion
- Defines the `appmodules` library, where all the app C++ code will be included.
- Loads the base React Native's CMake file
- Adds the Module C++ source code that we need to build with the `target_sources` directives. By default React Native will already populate the `appmodules` library with default sources, here we include our custom one. You can see that we need to crawl back from the `jni` folder to the `shared` folder where our C++ TM lives.
- Specify where CMake can find the module header files. Also in this case we need to crawl back from the `jni` folder.
```

> #### 2. Modify gradle.build to link the C++ code

Reviewer: ```suggestion
#### 2. Modify `build.gradle` to include the custom C++ code
```

> Gradle is the tool that orchestrates the Android build. We need to tell it where it can find the `CMake` files to build the Turbo Native Module.

Reviewer: ```suggestion
Gradle is the tool that orchestrates the Android build. We need to tell it where it can find the `CMake` files to build the Turbo Native Module.
```

1. Open the `SampleApp/android/app/build.gradle` file
2. Add the following block to the gradle file:

```diff
    if (hermesEnabled.toBoolean()) {
        implementation("com.facebook.react:hermes-android")
    } else {
        implementation jscFlavor
    }
}

+ android {
+   externalNativeBuild {
+       cmake {
+           path "src/main/jni/CMakeLists.txt"
+       }
+   }

> +}

Reviewer: I'd rather have something different here like:

```diff
android {
+   externalNativeBuild {
+       cmake {
+           path "src/main/jni/CMakeLists.txt"
+       }
+   }
}
```

as specifying 2x `android{}` block can have unintended side effects

```

> This block tells the Gradle file where to look for the external native build files. The path is relative to the folder where the gradle.build file lives, so we need to add the path to the `CMakeLists.txt` files in the `jni` folder.

Reviewer: ```suggestion
This block tells the Gradle file where to look for the CMake file. The path is relative to the folder where the `build.gradle` file lives, so we need to add the path to the `CMakeLists.txt` files in the `jni` folder.
```

```diff
#include <DefaultComponentsRegistry.h>
#include <DefaultTurboModuleManagerDelegate.h>
#include <autolinking.h>
#include <fbjni/fbjni.h>
#include <react/renderer/componentregistry/ComponentDescriptorProviderRegistry.h>
#include <rncore.h>

+ // Include the NativeSampleModule header
+ #include <NativeSampleModule.h>

//...

std::shared_ptr<TurboModule> cxxModuleProvider(
    const std::string& name,
    const std::shared_ptr<CallInvoker>& jsInvoker) {
  // Here you can provide your CXX Turbo Modules coming from
  // either your application or from external libraries. The approach to follow
  // is similar to the following (for a module called `NativeCxxModuleExample`):
  //
  // if (name == NativeCxxModuleExample::kModuleName) {
  //   return std::make_shared<NativeCxxModuleExample>(jsInvoker);
  // }

+  // This code register the module so that when the JS side asks for it, the app can return it
+  if (name == NativeSampleModule::kModuleName) {
+    return std::make_shared<NativeSampleModule>(jsInvoker);
+  }

  // And we fallback to the CXX module providers autolinked
  return autolinking_cxxModuleProvider(name, jsInvoker);
}

// leave the rest of the file
```

> These steps download the original OnLoad file from React Native, so that we can safely override it to load the C++ Turbo Native Module in the app.

Reviewer: ```suggestion
These steps download the original `OnLoad.cpp` file from React Native, so that we can safely override it to load the C++ Turbo Native Module in the app.
```

import SampleTurboModule from './specs/NativeSampleModule';

function App(): React.JSX.Element {
  const [value, setValue] = React.useState('');
  const [reversedValue, setReversedValue] = React.useState('');

  const onPress = () => {
    const revString = SampleTurboModule.reverseString(value);
    setReversedValue(revString);
  };

  return (
    <SafeAreaView style={styles.container}>
      <View>
        <Text style={styles.title}>
          Welcome to C++ Turbo Native Module Example
        </Text>
        <Text>Write down here he text you want to revert</Text>
        <TextInput
          style={styles.textInput}
          placeholder="Write your text here"
          onChangeText={setValue}
          value={value}
        />
        <Button title="Reverse" onPress={onPress} />
        <Text>Reversed text: {reversedValue}</Text>
      </View>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  title: {
    fontSize: 18,
    marginBottom: 20,
  },
  textInput: {
    borderColor: 'black',
    borderWidth: 1,
    borderRadius: 5,
    padding: 10,
    marginTop: 10,
  },
});

> 

Reviewer: nit: can we simplify this a bit?

Author: We could have an hardcoded string, but I feel like it's nice to provide users with something they can play with. :D
